node/src/client/mod.rs

```diff
@@ -181,6 +203,73 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
+                if node.shutdown.load(Relaxed) {
+                    info!("Shutting down transaction verification");
+                    break;
+                }
+
+                // Sleep briefly to allow transactions to be validated.
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+
+                // Determine the length of the queue.
+                let queue_len = node.transaction_queue.lock().len();
+
+                // If we don't have any transactions to verify, try again later.
+                if queue_len == 0 {
+                    continue;
+                }
+
+                // Determine how many transactions we want to verify.
+                let num_transactions = VERIFICATION_CONCURRENCY_LIMIT.min(queue_len);
+
+                // Check if we have room to verify the transactions and update the counter accordingly.
+                let previous_verification_counter = node.verification_counter.fetch_update(Relaxed, Relaxed, |c| {
+                    // If we are already verifying sufficient transactions, don't verify any more for now.
+                    if c >= VERIFICATION_CONCURRENCY_LIMIT {
+                        None
+                    // If we have space to verify more transactions, verify as many as we can.
+                    } else {
+                        // Consider verifying the full transaction queue, but limit to the concurrency limit.
+                        Some((c + queue_len).min(VERIFICATION_CONCURRENCY_LIMIT))
+                    }
+                });
+
+                // Determine how many transactions we cán verify.
+                let num_transactions = match previous_verification_counter {
+                    // Determine how many transactions we can verify.
+                    Ok(previous_value) => num_transactions.saturating_sub(previous_value),
+                    // If we are already verifying sufficient transactions, don't verify any more for now.
+                    Err(_) => continue,
+                };
+
+                // For each transaction, spawn a task to verify it.
+                let mut tx_queue = node.transaction_queue.lock();
+                for _ in 0..num_transactions {
+                    if let Some((_, (peer_ip, serialized, transaction))) = tx_queue.pop_lru() {
+                        let _node = node.clone();
+                        tokio::spawn(async move {
+                            // Check the transaction.
+                            if _node.ledger.check_transaction_basic(&transaction, None, &mut rand::thread_rng()).is_ok()
+                            {
+                                // Propagate the `UnconfirmedTransaction`.
+                                _node.propagate(Message::UnconfirmedTransaction(serialized), &[peer_ip]);
+                            }
+                            // Reduce the verification counter.
+                            _node.verification_counter.fetch_sub(1, Relaxed);
```

Comment:
this could be moved after the loop so that we update the atomic only once (subtracting `num_transactions`)

Reply:
I think that would no longer serve our goal, we only want the verification counter to be released when verification of a transaction is finished, not when we spawned all the verifications. Clarified the meaning a bit: 61e2a1d18

Reply:
ah, good point